tls/s2n_handshake_io.c

```diff
@@ -1211,11 +1216,34 @@ static int s2n_handshake_read_io(struct s2n_connection *conn)
         /* TLS1.3 can receive unexpected CCS messages at any point in the handshake
          * due to a peer operating in middlebox compatibility mode.
          * However, when operating in QUIC mode, S2N should not accept ANY CCS messages,
-         * including these unexpected ones.*/
+         * including these unexpected ones.
+         *= https://tools.ietf.org/rfc/rfc8446#5
+         *# An implementation may receive an unencrypted record of type
+         *# change_cipher_spec consisting of the single byte value 0x01 at any
+         *# time after the first ClientHello message has been sent or received
+         *# and before the peer's Finished message has been received and MUST
+         *# simply drop it without further processing.
+         */
         if (!IS_TLS13_HANDSHAKE(conn) || s2n_connection_is_quic_enabled(conn)) {
             POSIX_ENSURE(EXPECTED_RECORD_TYPE(conn) == TLS_CHANGE_CIPHER_SPEC, S2N_ERR_BAD_MESSAGE);
             POSIX_ENSURE(!CONNECTION_IS_WRITER(conn), S2N_ERR_BAD_MESSAGE);
         }
+        /* Check if we are after the handshake's finished message.
+         * Technicly we should also check that this isn't the very first message we
+         * recv as the server. Since we allow multiple CLIENT_HELLO's due to the
+         * retry request it is not obvious how to check this properly.
+         *= https://tools.ietf.org/rfc/rfc8446#5
+         *# If an implementation
+         *# detects a change_cipher_spec record received before the first
+         *# ClientHello message or after the peer's Finished message, it MUST be
+         *# treated as an unexpected record type (though stateless servers may
+         *# not be able to distinguish these cases from allowed cases).
+         */
+        if (IS_TLS13_HANDSHAKE(conn)) {
+            S2N_ERROR_IF(EXPECTED_RECORD_TYPE(conn) == TLS_APPLICATION_DATA, S2N_ERR_BAD_MESSAGE);
+            S2N_ERROR_IF(CONNECTION_WRITER(conn) == 'S' && !(conn->handshake.client_hello_seen), S2N_ERR_BAD_MESSAGE);
+            S2N_ERROR_IF(ACTIVE_MESSAGE(conn) == APPLICATION_DATA, S2N_ERR_BAD_MESSAGE);
+        }
```

Comment:
This isn't possible. We should never be executing this code for TLS_APPLICATION_DATA. That's s2n_recv's job.

Additionally, this part of the code only executes for TLS_HANDSHAKE messages.

Reply:
Oh and we don't use S2N_ERROR_IF anymore. Use POSIX_ENSURE.